_posts/2018-04-18-dependencies-done-right.md

```diff
@@ -0,0 +1,73 @@
+
+And that's the catch. When you specify a dependency, package managers are not required to give you the exact same version from anywhere in the dependency tree! Doing so is an optimization, and while we try to do our best to remove duplicated packages from the tree, it's not always feasible or possible. Because we cannot guarantee it, *you should not rely on us doing it*. By spec, we are allowed to change the behavior from a version to another if we think it might give a better result. We technically could even disable hoisting altogether!
+
+> **Why is it not guaranteed?** The first reason is that hoisting the packages is not always possible. Consider the following case: both Foo and Bar depend on HelloWorld@1.0.0, and the top-level package depends on Foo, Bar, and HelloWorld@2.0.0. Because of how the Node resolution works, Foo and Bar copies of HelloWorld cannot be hoisted (they would conflict with the version required by the top-level), meaning that the result of their `require()` calls will be different instances. 
+>
+> Another reason is that hoisting can be optimized for different things - do you want your dependency tree to use the latest possible versions - which possibly means a more stable application? Or do you want it to optimize for the size and merge all packages together as much as possible, even if it means not installing the latest patchs? In the end, your package managers usually have to follow an heuristic to make their decisions, which might sometimes lead to accepting different tradoffs from what you would originally expect.
+
+## Making Things Right
+
+Since we now know that using the regular `dependencies` field won't always merge our dependencies with similar ones used by others packages, what shall we use instead? The short answer is **peer dependencies**.
+
+See, peer dependencies have this particular property to ensure the dependencies of your package listed here should be the exact same ones with the ones used by your parent package in the dependency tree. Even better, they also guarantee that you will always get the *exact* same instance with them, even if we disable the hoisting! This is literally why they exist.
+
+So without further ado, this is what we should have done:
+
+```json
+{
+    "name": "my-awesome-plugin",
+    "peerDependencies": {
+        "react": "^16.0.0"
+    }
+}
+```
+
+Of course, there is a slight inconvenience: our users will now have to add `react` to their own dependencies if they haven't done so already (). If they don't, then they'll get a warning, and we won't be able to access it.
```

Comment:
Why do we have the empty parenthesis here: `if they haven't done so already ()`

Reply:
Which parentheses? I see no parentheses! 😄